libcloud/dns/drivers/digitalocean.py

```diff
@@ -0,0 +1,291 @@
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Digital Ocean DNS Driver
+"""
+
+__all__ = [
+    'DigitalOceanDNSDriver'
+]
+
+from libcloud.utils.py3 import httplib
+
+from libcloud.common.digitalocean import DigitalOcean_v2_BaseDriver
+from libcloud.common.digitalocean import DigitalOcean_v2_Connection
+from libcloud.dns.types import Provider, RecordType
+from libcloud.dns.base import DNSDriver, Zone, Record
+
+
+class DigitalOceanDNSDriver(DigitalOcean_v2_BaseDriver, DNSDriver):
+    connectionCls = DigitalOcean_v2_Connection
+    type = Provider.DIGITAL_OCEAN
+    name = "DigitalOcean"
+    website = 'https://www.digitalocean.com'
+
+    RECORD_TYPE_MAP = {
+        RecordType.NS: 'NS',
+        RecordType.A: 'A',
+        RecordType.AAAA: 'AAAA',
+        RecordType.CNAME: 'CNAME',
+        RecordType.MX: 'MX',
+        RecordType.TXT: 'TXT',
+        RecordType.SRV: 'SRV',
+    }
+
+    def list_zones(self):
+        """
+        Return a list of zones.
+
+        :return: ``list`` of :class:`Zone`
+        """
+        data = self._paginated_request('/v2/domains', 'domains')
+        return list(map(self._to_zone, data))
+
+    def list_records(self, zone):
+        """
+        Return a list of records for the provided zone.
+
+        :param zone: Zone to list records for.
+        :type zone: :class:`Zone`
+
+        :return: ``list`` of :class:`Record`
+        """
+        data = self._paginated_request('/v2/domains/%s/records' % (zone.id),
+                                       'domain_records')
+# TODO: Not use list comprehension to add zone to record for proper data map
+#       functionality?
+        return list(map(self._to_record, data, [zone for z in data]))
```

Comment:
This is a tricky issue. It's reflected in the comment above it. So, the main action here is mapping the records ( list in data ) to the zone. In this particular case, I am building a list of the same zone to match the size of the list in data so that map can do it's thing.

So in this case, _to_record is actually getting an item from the `data` list, and an item from the `zone` list ( which is the same zone for all items in data ) through the map function so that it spits out a proper list of Record objects.

The reason I don't use the named arguments, is because this is in map which will break since it doesn't use named arguments.

```Traceback (most recent call last):
  File "libcloud/test/dns/test_digitalocean.py", line 48, in test_list_records
    records = self.driver.list_records(zone)
  File "/home/jcastillo2nd/Development/libcloud/libcloud/dns/drivers/digitalocean.py", line 69, in list_records
    return list(map(self._to_record, data=data, zone=[zone for z in data]))
TypeError: map() takes no keyword arguments```

I would be glad to use the keyword here, but to have it work, and until a proper implementation can be generated for mapping this without making the ephemeral list of `zone` to properly get mapped to the `data`, I am at a loss. Any ideas on this would be appreciated on this, as this has bugged me from the beginning and though my memory is fuzzy, I think it was a compromise I made to generate proper record objects with real references to the Zone.

In the interim, I've gone ahead and added to the comment to clarify the lack of keyword usage here for map, and more clarification on why that list generator is used.